Changes registry test helper to support non-localhost

micahyoung opened the pull request:
This is useful for consumers of DockerRegistry helper, particularly lifecycle (maybe pack, whose [impl](https://github.com/buildpacks/pack/blob/main/testhelpers/registry.go#L182) this was based on and already supports this), when running test containers that access the separate registry container, listening on a published host port. Currently, this was only possible with `--network=host` which can't be used on Windows. This also lets a Unix workstation run tests against a remote Windows (or any arch/OS) daemon with `DOCKER_HOST=tcp://<daemon host>` set.

The test registry now has a `Host` field, whose value is autodetected from the host and which can be used in image refs that need to be written to it: ex: `registry.Host + ":" + registry.Port + "/image-name:tag"`.

The logic to detect the host IP:
  - Any exact IPs (`<host IP>/32`) in `insecure-registry` entries, assume that's the host's IP
  - ... or, fallback to "localhost"

This does not change the current behavior for imgutil when running on GHA, which has none of the above set so will use "localhost".

This also replaces `stefanscherer/registry-windows`, which had a bug where it would return sporadic 503s. This bug appears fixed in docker-distribution 2.7.1 but the image hasn't been update for over 2 years. This is replaced with a newly created [`micahyoung/registry`](https://github.com/micahyoung/registry-image) which hosts equivalent Windows & Linux AMD64/ARM images in a single manifest list and is updated automatically through GHA. This should make the test registry behave as consistently as possible, regardless of the OS/Arch. I'd be happy to donate this to `buildpacks` org or even move the `Dockerfile`s into this repo to be built on-demand.

Signed-off-by: Micah Young <ymicah@vmware.com>

(updated, remove extra registry host detection logic for simplicity)

micahyoung commented:
For reference, lifecycle Draft PR (with analyzer tests enabled for Windows): https://github.com/buildpacks/lifecycle/pull/541

ekcasey commented:
Looks good, but I do worry a touch about making it so easy to open up a totally unauthenticated registry on a public network. The analyzer tests specifically tests scenarios when auth is not required to pull images. Could we do either or both
* Make the registry created with `NewDockerRegistry` require auth to write but not to read?
* Rename `NewDockerRegistry` to something like `NewDockerRegistryWithoutAuth` (or `NewDockerRegistryWithPublicRead` if we do the bullet point above) and add a doc comment telling folks to be careful if/when they use this that they have other network-level protections in place.

micahyoung commented:
@ekcasey Great points. I'll take a pass a public-read/authenticated-write as the overridable defaults (with very explicit warnings).

micahyoung commented:
I updated this with the readwrite-with-auth/readonly-without-auth approach and added some (hopefully) useful test coverage to ensure these get helpers get exercised. I added registry helper options for auth and shared-volumes.  I bumped https://github.com/buildpacks/lifecycle/pull/541 to show how the dual-registry-shared-data works and everything is green.

natalieparellano commented on testhelpers/testhelpers.go:
```diff
@@ -177,8 +179,25 @@ func DockerRmi(dockerCli dockercli.CommonAPIClient, repoNames ...string) error {
+	ref, err := name.ParseReference(refStr, name.WeakValidation)
+	if err != nil {
+		return err
+	}
```
```suggestion
	}
	
```

natalieparellano commented on testhelpers/docker_registry.go:
```diff
@@ -103,34 +159,85 @@ func (r *DockerRegistry) Start(t *testing.T) {
 	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"username":"%s","password":"%s"}`, r.username, r.password)))
 }
 
+func DockerHostname(t *testing.T) string {
```
Would it be worth adding a docs comment here describing the chain of precedence?

natalieparellano commented on testhelpers/docker_registry.go:
```diff
@@ -73,17 +95,51 @@ func (r *DockerRegistry) Start(t *testing.T) {
+	var volumeBinds []string
+	var containerUser string
+	if r.volumeName != "" {
+		// try to create volumes that do not exist
```
```suggestion
		// try to create volumes that may exist
```

natalieparellano commented on remote/remote_test.go:
```diff
@@ -1027,6 +1042,33 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
+
+				h.AssertNil(t, img.Save())
+
+				name, err := name.ParseReference(repoName, name.WeakValidation)
```
I'm a bit confused... how does this work? It looks like `repoName` is on the authenticated registry. Where are we pulling from the unauthenticated one?

micahyoung replied:
Yeah, this could use some clarification. What it attempts to do is:
1. Write a new image on the auth'd registry with ref like: `localhost:5555/my-image-1234`
2. Parse out the name out of the auth'd ref: `my-image.1234`
3. Rebuild the ref string for the same image on the unauth'd repo: `localhost:6666/my-image-1234`
4. Show that we can pull that without any credentials (since we never added any for `localhost:6666` to the `DOCKER_CONFIG`).

micahyoung replied:
I'll take another pass and see if I can clean it up. TBH, the test is mostly here for some coverage for our `WithSharedStorageVolume` helper, theoretically unauthenticated registries are covered with the DockerHub examples.

ekcasey commented on testhelpers/testhelpers.go:
```diff
@@ -177,8 +179,26 @@ func DockerRmi(dockerCli dockercli.CommonAPIClient, repoNames ...string) error {
 
-func PushImage(dockerCli dockercli.CommonAPIClient, ref string) error {
-	rc, err := dockerCli.ImagePush(context.Background(), ref, dockertypes.ImagePushOptions{RegistryAuth: "{}"})
+//PushImage pushes and image to a registry, optionally using credentials from any set DOCKER_CONFIG
```
```suggestion
//PushImage pushes an image to a registry, optionally using credentials from any set DOCKER_CONFIG
```

ekcasey reviewed (approved):
This looks fantastic!
